### PR TITLE
🎲 [GRPO] Make training dataset shuffle optional

### LINKED ISSUE #2998: GRPOTrainer does not have a feature flag to prevent dataset shuffling
### Reproduction

In [GRPOTrainer](https://github.com/huggingface/trl/blob/491921c1a4167e7c84429382470b0bb3158e66b0/trl/trainer/grpo_trainer.py#L144):

```python
def __iter__(self):
    # E.g., [2, 4, 3, 1, 0, 6, 5] (num_samples = 7)
    indexes = torch.randperm(self.num_samples, generator=self.generator).tolist()

    #    [2, 4, 3, 1, 0, 6, 5]
    # -> [[2, 4, 3], [1, 0, 6], [5]]  (batch_size = 3)
    indexes = [indexes[i : i + self.batch_size] for i in range(0, len(indexes), self.batch_size)]

    #    [[2, 4, 3], [1, 0, 6], [5]]
    # -> [[2, 4, 3], [1, 0, 6]]
    indexes = [chunk for chunk in indexes if len(chunk) == self.batch_size]

    for chunk in indexes:
        for _ in range(self.repeat_count):
            for index in chunk:
                for _ in range(self.mini_repeat_count):
                    yield index

def __len__(self) -> int:
    return self.num_samples * self.mini_repeat_count * self.repeat_count
```

outputs:

Shuffled dataset even though I want to apply Curriculum Learning (no shuffling). I've pasted some code that works the way I want below, but would love any advice/help in potentially creating a PR for this issue since this would be my first time contributing, and I am unsure what other files would need to be changed (i.e. GRPOConfig, PretrainedConfig, etc.). I also want to keep the option for shuffling based on the Config class since others will likely want to retain the original behavior.

### Potential Solution:

Modify or create new class similar to `RepeatRandomShuffler` called `RepeatSequentialShuffler` that modifies the iteration logic as follow:

```python
def __iter__(self):
    # Sequential order, repeat each index `repeat_count` times
    indexes = [idx for idx in range(self.num_samples) for _ in range(self.repeat_count)]
    return iter(indexes)

def __len__(self):
    return self.num_samples * self.repeat_count
```

### System Info

- Platform: Linux-5.15.0-113-generic-x86_64-with-glibc2.35
- Python version: 3.12.9
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA A100-SXM4-80GB
- Transformers version: 4.47.1
- Accelerate version: 1.2.0
- Accelerate config: not found
- Datasets version: 3.1.0
- HF Hub version: 0.29.1
- TRL version: 0.12.1
- bitsandbytes version: 0.45.0
- DeepSpeed version: 0.16.1
- Diffusers version: 0.32.2
- Liger-Kernel version: 0.4.2
- LLM-Blender version: not installed
- OpenAI version: 1.64.0
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by LeonEricsson**
Fixes #2998, #3333 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/test_grpo_trainer.py [hunk 1]
```diff
@@ -46,6 +46,21 @@ def test_sampler(self):
         # Check that each element is repeated twice
         assert all(sampled[i] == sampled[i + 1] for i in range(0, len(sampled), 2))
 
+    def test_sampler_noshuffle(self):
+        dataset = ["a", "b", "c", "d", "e", "f", "g"]
+        sampler = RepeatRandomSampler(dataset, mini_repeat_count=2)
+        # Should output something like [4, 4, 3, 3, 0, 0, 1, 1, 2, 2, 6, 6, 5, 5]
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py:52]:
```suggestion
        # Should output [0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6]
```

File: tests/test_grpo_trainer.py [hunk 2]
```diff
@@ -46,6 +46,21 @@ def test_sampler(self):
...
+        sampled = list(sampler)
+        # Check that the length is doubled
+        assert len(sampled) == 2 * len(dataset)
+        # Check that all indexes are present
+        assert set(sampled) == set(range(len(dataset)))
+        # Check that each element is repeated twice
+        assert all(sampled[i] == sampled[i + 1] for i in range(0, len(sampled), 2))
+
+        # new
+        assert sampled == [0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6]
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py:62]:
```suggestion
        self.assertEqual(sampled, [0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6] )   
```

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -78,7 +78,7 @@
 RewardFunc = Union[str, PreTrainedModel, Callable[[list, list], list[float]]]
 
 
-class RepeatRandomSampler(Sampler):
+class RepeatSampler(Sampler):
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:81]: This renaming makes sense. But to ensure backward compatibility, can you just add something like

```python
class RepeatRandomSampler(RepeatSampler):
    def __init__(self, *args, **kwargs):
        warnings.warn(
            "RepeatRandomSampler is deprecated and will be removed in version 0.18. Use RepeatSampler instead",
            DeprecationWarning,
        )
        super().__init__(*args, **kwargs)
```

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:81]: And you also need to rename in `_get_train_sampler`

File: tests/test_grpo_trainer.py (hunk 2 above)
Comment (qgallouedec) [reply tests/test_grpo_trainer.py]: Sorry about that! I was curious to check your work 😅

File: tests/test_grpo_trainer.py [hunk 4]
```diff
@@ -46,93 +46,128 @@ def test_sampler(self):
...
-        assert len(sampled) == 6 * (len(dataset) - 1)  # 1 element is dropped, because it's not enough to form a batch
-        # Check that the sampled indexes are a subset of the dataset indexes
-        assert set(sampled).issubset(set(range(len(dataset))))
-        # Check that each element is repeated as expected
-        assert all(sampled[i] == sampled[i + 1] == sampled[i + 2] for i in range(0, len(sampled), 3))
-        # Check that the batch is repeated as expected
-        assert sampled[0:6] == sampled[6:12]
-        assert sampled[12:18] == sampled[18:24]
-        assert sampled[24:30] == sampled[30:36]
+        expected = [
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py:88]: any recommendation on how to avoid ruff reformatting this? I think you can do something like

```
expected = (
    [0, 0, 0, 1, 1, 1] +
    [0, 0, 0, 1, 1, 1] +
    [2, 2, 2, 3, 3, 3] +
    [2, 2, 2, 3, 3, 3] +
    [4, 4, 4, 5, 5, 5] +
    [4, 4, 4, 5, 5, 5]
)
```

File: tests/test_grpo_trainer.py
Comment (qgallouedec) [inline tests/test_grpo_trainer.py:1]: I think there is no need to modify all the tests. Just adding a new test for the case `shuffle=False` is enough

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -166,6 +175,15 @@ def __len__(self) -> int:
         return self.num_samples * self.mini_repeat_count * self.repeat_count
 
 
+class RepeatRandomSampler(RepeatSampler):
+    def __init__(self, *args, **kwargs):
+        warnings.warn(
+            "RepeatRandomSampler is deprecated and will be removed in version 0.18. Use RepeatSampler instead",
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:181]:
```suggestion
            "RepeatRandomSampler is deprecated and will be removed in version 0.18. Use RepeatSampler instead.",
```

**Review by qgallouedec — COMMENTED** [review]
Just some minor things, and I think we're mostly good!

File: tests/test_grpo_trainer.py
Comment (LeonEricsson) [reply tests/test_grpo_trainer.py:1]: Alright! although I do feel the clarity/readability improves quite a bit

**Comment by Maghoumi** [comment]
@LeonEricsson 

Does this actually work? Maybe there's a trick to getting this to work that I'm not aware of?

On my end, I sort my training dataset based on some criteria, then I set the `shuffle_dataset=False` when instantiating the `GRPOTrainer`. Also, I confirm that the dataset that is being passed to `GRPOTrainer`'s constructor still has the order I sorted in.

But when the rewards are calculated, I see that the order of the inputs passed to my reward function is random as if it's shuffled.

**Comment by LeonEricsson** [comment]
> Does this actually work? Maybe there's a trick to getting this to work that I'm not aware of?
> 
> On my end, I sort my training dataset based on some criteria, then I set the `shuffle_dataset=False` when instantiating the `GRPOTrainer`. Also, I confirm that the dataset that is being passed to `GRPOTrainer`'s constructor still has the order I sorted in.
> 
> But when the rewards are calculated, I see that the order of the inputs passed to my reward function is random as if it's shuffled.

Created a quick sanity script and things worked fine for me. The prompts in the reward function appeared in the expected sorted order

```python
import random

from datasets import load_dataset

from trl import GRPOConfig, GRPOTrainer


# Load and sort dataset by a predictable field
dataset = load_dataset("trl-lib/tldr", split="train[:10]")

# sory by prompt length
prompt_lengths = [len(item["prompt"]) for item in dataset]
dataset = dataset.add_column("prompt_length", prompt_lengths)
dataset = dataset.sort("prompt_length", reverse=False)


def reward_random(prompts, completions, **kwargs):
    for p in prompts:
        print(len(p))
    return [random.random() for _ in range(len(completions))]


trainer = GRPOTrainer(
    model="Qwen/Qwen2.5-0.5B-Instruct",
    reward_funcs=reward_random,
    train_dataset=dataset,
    args=GRPOConfig(
        per_device_train_batch_size=2,
        num_generations=2,
        max_completion_length=128,
        max_steps=10,
        report_to="none",
        shuffle_dataset=False,
    ),
)

trainer.train()
```

**Comment by Maghoumi** [comment]
Thanks for your response and sharing your minimal code to reproduce. Your code helped me in two different ways:

1. I was passing `shuffle_dataset` to `GRPOTrainer` (instead of passing it to `GRPOConfig`). Duh! I confirm it's working as expected once passed correctly.
2. I detected a strange behavior in unsloth, where if you run your code on a machine with multiple GPUs (and all the GPUs are visible to the code), the behavior of the `GRPOTrainer`'s sampler is different than if you run it with only a single GPU (or if only a single GPU is visible via `CUDA_VISIBLE_DEVICES`). This is despite the fact that unsloth only uses a single GPU for training. So, somehow the number of GPUs affect the batching logic. I will follow up with them on this.

Thanks again for your help.

**Comment by qgallouedec** [comment]
I don't think it's unsloth but transformers' Trainer. When you have multiple GPU visible, you must use `CUDA_VISIBLE_DEVICES`

**Comment by LeonEricsson** [comment]
> I don't think it's unsloth but transformers' Trainer. When you have multiple GPU visible, you must use `CUDA_VISIBLE_DEVICES`

yeah i've experienced this as well
